Hotel 7

dudarenko-dmitry opened the pull request:
Task_7_Hotel is completed:

Task 1 (difficulty 3)
Add the ability to configure the program from the previous task using the property file:
- Enabling/disabling the ability to change the status of a number;
- Number of guest records in the room history.

Task 2 (difficulty 5)
Implement program state preservation by serializing its objects to a file. Restore the saved state at the start, and write the new state to a file at the end.
It implies the use of standard serialization mechanisms (Serializable or Externalizable), but it is permissible to use other ones (for example, Jackson as a serializer in JSON).

dudarenko-dmitry committed:
Task_7:
1. Created branch Hotel_7
2. Edited README.md

dudarenko-dmitry committed:
Task_7:
Created AppPropertiesLoader and AppConfiguration

dudarenko-dmitry committed:
Task_7:
Updated logic in ReaderWriter:
1. Created EntitiesConverter and move here all logic for getting path, headers and converting.
2. Left in ReaderWriter logic only for Read and Write.

dudarenko-dmitry committed:
Task_7:
Updated logic in ReaderWriter:
1. Created EntitiesConverter and move here all logic for getting path, headers and converting.
2. Left in ReaderWriter logic only for Read and Write.

dudarenko-dmitry committed:
Task_7:
Updated logic in ReaderWriter:

dudarenko-dmitry committed:
Merge remote-tracking branch 'origin/Hotel_7' into Hotel_7

dudarenko-dmitry committed:
Task_7:
Created methods for changing RoomStatus in Room.
Added check of Property "ValueIsAbleToChangeRoomStatus" in ServiceRoom.
 Added checking of RoomStatus in during RoomReservation's creation.

dudarenko-dmitry committed:
Task_7:
Created methods for changing RoomStatus in Room.
Added check of Property "ValueIsAbleToChangeRoomStatus" in ServiceRoom.
Added checking of RoomStatus in during RoomReservation's creation.

Added check of Property "NUMBER_OF_GUEST_RECORDS_FOR_ROOM" in ServiceRoomReservationImpl.
Added checking of NumberOfRecords in during RoomReservation's creation.

Added reading of files' names for ReaderWriter from Properties.

Created 2 ways of loading/saving Applications' data:
1. using DataProcessorFile to control loading/saving of all entities during start and exit of/from Application and during runtime.
2. using ProcessorSerializable for Serialization of Entity Hotel where joined all data lists.

Added Properties for Serialization.

dudarenko-dmitry committed:
Task_7:
Added check of Property "NUMBER_OF_GUEST_RECORDS_FOR_ROOM" in ServiceRoomReservationImpl.
 Added checking of NumberOfRecords in during RoomReservation's creation.

dudarenko-dmitry committed:
Task_7:
Added reading of files' names for ReaderWriter from Properties.

dudarenko-dmitry committed:
Task_7:
Created DataProcessorFile to control loading/saving of all entities during start and exit of/from Application and during runtime.

dudarenko-dmitry committed:
Task_7:
Created 2 ways of loading/saving Applications' data:
1. using DataProcessorFile to control loading/saving of all entities during start and exit of/from Application and during runtime.
2. using ProcessorSerializable for Serialization of Entity Hotel where joined all data lists.

dudarenko-dmitry committed:
Task_7:
added Properties for Serialization.

dudarenko-dmitry committed:
Merge remote-tracking branch 'origin/Hotel_7' into Hotel_7

Mark005 commented on src/pl/senla/hotel/configuration/AppConfiguration.java:
```diff
@@ -0,0 +1,63 @@
+    public static final String KEY_FILE_ORDERS_NAME = "File name for Orders";
+
+    private final PropertiesLoader propertiesLoader = new AppPropertiesLoader();
+    private final Map<String, String> properties = propertiesLoader.loadConfiguration();
```
move this to the constructor, please

dudarenko-dmitry replied:
Like this?  Right?
   
   private final Map<String, String> properties;
   
   public AppConfiguration() {
        PropertiesLoader propertiesLoader = new AppPropertiesLoader();
        properties = propertiesLoader.loadConfiguration();
}

Mark005 replied:
yes, almost

Mark005 replied:
propertiesLoader = new AppPropertiesLoader();
properties = propertiesLoader.loadConfiguration();

Mark005 commented on src/pl/senla/hotel/configuration/AppPropertiesLoader.java:
```diff
@@ -0,0 +1,71 @@
+        Map<String, String> configuration = new HashMap<>();
+        try (FileInputStream fis = new FileInputStream(propertiesFile)) {
+            appProperties.load(fis);
+            configuration.put(KEY_ABLE_TO_CHANGE_ROOM_STATUS,
```
you shouldn't put anything in properties while the program is running, so the property file should be readable only for the application), but you should prepare this file before running the application

dudarenko-dmitry replied:
I did exactly the same:
1. I prepare prorties file in separate method, but this method is called in Main. This method is called only the first time, so now I commented it in Main.
2. I read properties during starting of application and before reading all data from saved files.

Is it right way of using properties-file?

dudarenko-dmitry replied:
all loaded properties I put to Map and later use values from Map (receive by keys) in neccesery places.

Could I do this by this  way  or will be better implementation without Map and read values directly from properties-file?

mrseneca replied:
This and the previous comment about the constructor should be considered together.

You can read your file and prepare all the properties for use immediately when starting the program, because it is expected that the file cannot be changed while the program is running.

Therefore, it is better to transfer the file reading logic to the constructor of the class that will operate with them, i.e. has Properties.java.

Additionally, it seems to me that it is unnecessary to create an additional map for properties, the Properties field itself is enough. If you look inside this class, you can see that they are already stored in the map, but with a lot of ready-made functionality with checks and receiving.

It is enough to simply load the values into Properies during the execution of the constructor, and make methods for obtaining values by properties names, for example:

```
public Integer getIntProperty(String propertyname) {
return Integer.parseInt(appProperties.getProperty(propertyname));
}

public Boolean getIntProperty(String propertyname) {
return Boolean.parseBoolean(appProperties.getProperty(propertyname));
}
```

mrseneca replied:
Also I can't find you properties file. Did you pushed it into git-repo?

dudarenko-dmitry replied:
> Also I can't find you properties file. Did you pushed it into git-repo?

my mistake: saved it in separate directory. Now added in my project  a new package Resources, put hotel.properties there and pushed to repo.

dudarenko-dmitry replied:
> 
> Additionally, it seems to me that it is unnecessary to create an additional map for properties, the Properties field itself is enough. If you look inside this class, you can see that they are already stored in the map, but with a lot of ready-made functionality with checks and receiving.
> 

I agree with it.
I rewrote without Map.

dudarenko-dmitry replied:
> It is enough to simply load the values into Properies during the execution of the constructor, and make methods for obtaining values by properties names, for example:

It has been implemented here:  pl.senla.hotel.configuration.AppConfiguration.
I've just edited it for using Properties (instead of unnecessary using of Map). And make it as a singleton to keep all properties data which were loaded while its first loding during the start of program.

dudarenko-dmitry replied:
ready.
I hope I've done it correctly.

Mark005 commented on src/pl/senla/hotel/ie/file/EntitiesConverter.java:
```diff
@@ -12,74 +13,34 @@
 import static pl.senla.hotel.constant.InputOutputConstant.*;
 
-public class ReaderWriter<T> implements ReaderWriterUniversal<T>{
+public class EntitiesConverter<T> implements Converter<T> {
```
absolutely  wrong implementation, you should apply converter as a parameter for ReaderWriter but not implement it, so you will have some converters which will focused only on one class, so you will provide one of them in ReaderWriter and it will work with it

dudarenko-dmitry replied:
I'm sorry, Mark. probably I didn't catch your comment.
My ReaderWriter looked so:
public class ReaderWriter<T> implements ReaderWriterUniversal{
    private final Converter<T> converter;
    public ReaderWriter() {
        this.converter = new EntitiesConverter<>();
    }

Should I add     "Converter<T> converter"    as a parametr to constructor?
public ReaderWriter(Converter<T> converter) {
        this.converter = new EntitiesConverter<>();
    }

Did I understand correctly?

mrseneca replied:
The idea is that the reader only reads your CSV file, transfers the content to the converter, and it converts it into the desired entity and returns it.
With this approach, we can easily comply with the SOLID principles and get rid of code duplication (or large switch/if blocks) using abstract classes.

Yes, you must pass a certain converter to the reader, for certain entities, this can be done through a constructor.

dudarenko-dmitry replied:
ready.

Mark005 reviewed (changes requested):
please read it as example https://mkyong.com/java/java-properties-file-examples/

dudarenko-dmitry committed:
Task_9:
made adjustments to the code:
1. move name of Properties' keys to Constants.
2. use in SaveConfiguration appProperties.setProperty instead of appProperties.put.
3. add Converter as a parameter in ReaderWriter.
4. add to abstract class HotelService implementation of Serializable.

dudarenko-dmitry committed:
Task_7 :
made adjustments to the code:
1. move name of Properties' keys to Constants.
2. use in SaveConfiguration appProperties.setProperty instead of appProperties.put.
3. add Converter as a parameter in ReaderWriter.
4. add to abstract class HotelService implementation of Serializable.

dudarenko-dmitry committed:
Merge remote-tracking branch 'origin/Hotel_7' into Hotel_7

dudarenko-dmitry committed:
Task_7:
1. Moved hotel.properties from outside into package RESOURCES.
2. Made AppConfiguration as a Singleton.
3. Added reading of application's configuration from hotel.properties in Main while start.
4. Added loaded configuration to all necessary constructors.

dudarenko-dmitry committed:
Task_7:
Refactored:
1. Getting some properties.
2. Using of Generics in ReaderWriter and Converter.

dudarenko-dmitry committed:
Task_7:
Resolved several issues with import-export and serialization

dudarenko-dmitry commented:
Hi. 
please check one more time.

mrseneca commented on src/pl/senla/hotel/configuration/AppConfiguration.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    @Override
+    public Boolean getValueIsAbleToChangeRoomStatus() {
```
You did not need separate method to each property. In long-term development, the number of properties can exceed hundreds. In this case, your class will be huge. You just need to write general methods for each type passing the name of the property, as I already suggested in some comment

dudarenko-dmitry replied:
> You did not need separate method to each property. In long-term development, the number of properties can exceed hundreds. In this case, your class will be huge. You just need to write general methods for each type passing the name of the property, as I already suggested in some comment

Aleh, should I use this in appropriate places in program?
    pl.senla.hotel.service.ServiceRoomImpl (for example here)
    Boolean.valueOf(configuration.getProperties().getProperty(KEY_ABLE_TO_CHANGE_ROOM_STATUS))

instead of reading of each property by it's key in the same places using get-methods for this property from properties-file which I got in Main: 
    configuration.getValueIsAbleToChangeRoomStatus()
 
Did I understand correctly?

dudarenko-dmitry replied:
![image](https://github.com/dudarenko-dmitry/senla_training/assets/102149674/97b8efd5-98ff-4602-ad8e-7566311644a2)

mrseneca replied:
Not certainly in that way. Your property class should not only store your properties, but also give them in a convenient form so that you don’t have to get them through the internal fields of the object. Ideally the usage should look something like this:

```
if(configuration.getBooleanProperty(KEY_ABLE_TO_CHANGE_ROOM_STATUS)) {
   ...
}
```

mrseneca replied:
And property class in that case only need few methods

![изображение](https://github.com/dudarenko-dmitry/senla_training/assets/31473268/e674cb5f-21d9-4f57-a119-fee403567bb2)

dudarenko-dmitry replied:
... and I should implement such changes:
![image](https://github.com/dudarenko-dmitry/senla_training/assets/102149674/6c77af2f-3931-42b2-b14c-fee76e075bfc)

dudarenko-dmitry replied:
I got it.
Everithing will be fixed in 10 min

mrseneca commented on src/pl/senla/hotel/configuration/AppPropertiesLoader.java:
```diff
@@ -0,0 +1,68 @@
+    public Properties loadConfiguration() {
+        try (FileInputStream fis = new FileInputStream(propertiesFile)) {
+            appProperties.load(fis);
+            appProperties.put(KEY_ABLE_TO_CHANGE_ROOM_STATUS,
```
Um... Why is this work even happening? Do I understand correctly that you simply put in the properties the same value that you take from the same properties?

dudarenko-dmitry replied:
agree... 
I used them with previous version with unnecessary Map.
Without Map i all lines with "put" should be deleted

mrseneca commented on src/pl/senla/hotel/configuration/AppPropertiesLoader.java:
```diff
@@ -0,0 +1,68 @@
+     */
+    @Override
+    public void saveConfiguration() {
+        try (FileOutputStream fos = new FileOutputStream(propertiesFile)) {
```
I never saw in some project what properties was saved from application. Usually it's one way road

Also, i thing you can easily combine your AppPropertiesLoader  and AppConfiguration in one class. Coz they are both working with properties and have a strong connection, in fact they cannot be used alone without one

dudarenko-dmitry replied:
I create this method and use it only during the first start of application in Main to write properties-file correctly.
After creating of file tjis method should be deleted/commented as I've written in Main:
        // use to save new Properties
        PropertiesLoader loader = new AppPropertiesLoader();
        loader.saveConfiguration();

Mark005 replied:
you should only read properties from file, so it should be there BEFORE application starting, remove please all **setProperty**

mrseneca commented on src/pl/senla/hotel/resources/hotel.properties:
```diff
@@ -0,0 +1,10 @@
+#Fri Sep 22 00:40:29 TRT 2023
+permissionToChangeRoomStatus=TRUE
```
usually properties have kebab-case style with dot as a separator for logical splt:
```
room-change.enabled=true
hotel-facilities.file-name=HotelFacility.csv
room-history.records.count=10
```

Try to avoid "to", "a", "for", etc.

dudarenko-dmitry replied:
edited

mrseneca commented on src/pl/senla/hotel/ie/file/EntitiesConverter.java:
```diff
@@ -109,13 +71,15 @@ private String[] convertEntityToString(T t) {
     }
 
-    private T convertStringToEntity(Class<? super T> clazz, String csvT) {
+    public T convertStringToEntity(Class<T> clazz, String csvT) {
```
I don't like this huge if statements. It will grow even bigger if new entities will added to application. 
Better to create different implementations of converters for different entities and pass them into reader method to convertation or something like that.

But wait for Mark's review also

dudarenko-dmitry replied:
create 4 different converters again and pass them to ReaderWriters' constructors.

mrseneca reviewed (changes requested):
Needs small fix

dudarenko-dmitry committed:
Task_7:
1. edited properties names.
2. deleted unnecessary lines in:
public Properties loadConfiguration()

dudarenko-dmitry committed:
Task_7:
Edited logic of Reading and Writing of Entities:
1. create 4 converters for every entity.
2. every of these converters are used in ReaderWriterEntity's constructors.
3. ReaderWriterEntities are used in DataProcessorFileEntity's constructor.

old classes moved to archive-package.

dudarenko-dmitry commented:
> Needs small fix

fixed.
but there are some questions.

dudarenko-dmitry committed:
Task_7:
refactored reading of Properties.

Mark005 commented on src/pl/senla/hotel/ie/file/ReaderWriterEntity.java:
```diff
@@ -0,0 +1,57 @@
+
+
+    @Override
+    public List<T> load(Class<T> clazz) throws IOException {
```
clazz is odd here because you have generic in class definition

dudarenko-dmitry replied:
thank you.

Mark005 commented on src/pl/senla/hotel/ie/file/ReaderWriterEntity.java:
```diff
@@ -0,0 +1,57 @@
+
+import static pl.senla.hotel.constant.InputOutputConstant.DATA_SAVED;
+
+public class ReaderWriterEntity<T> implements ReaderWriterUniversalEntity<T> {
```
nice job, now it looks great, that's how a real OOP approach should looks like

dudarenko-dmitry replied:
thank you.

Mark005 commented on src/pl/senla/hotel/ie/file/ReaderWriterEntity.java:
```diff
@@ -0,0 +1,57 @@
+
+import static pl.senla.hotel.constant.InputOutputConstant.DATA_SAVED;
+
+public class ReaderWriterEntity<T> implements ReaderWriterUniversalEntity<T> {
```
remove ReaderWriterUniversalEntity, now it should works without it

Mark005 commented on src/pl/senla/hotel/ie/file/archive/Converter.java:
```diff
@@ -0,0 +1,9 @@
+    String getPath(Class<T> clazz);
+    String[] getHeader(Class<T> clazz);
+    String[] convertEntityToString(T t);
+    T convertStringToEntity(Class<T> clazz, String csvT);
```
you don't need clazz here

Mark005 commented on src/pl/senla/hotel/ie/file/archive/Converter.java:
```diff
@@ -0,0 +1,9 @@
+
+public interface Converter<T> {
+
+    String getPath(Class<T> clazz);
```
you don't need clazz here

Mark005 commented on src/pl/senla/hotel/ie/file/archive/Converter.java:
```diff
@@ -0,0 +1,9 @@
+public interface Converter<T> {
+
+    String getPath(Class<T> clazz);
+    String[] getHeader(Class<T> clazz);
```
you don't need clazz here

Mark005 commented on src/pl/senla/hotel/ie/file/archive/EntitiesConverter.java:
```diff
@@ -12,74 +13,36 @@
+import static pl.senla.hotel.constant.PropertiesConstant.*;
 
-public class ReaderWriter<T> implements ReaderWriterUniversal<T>{
+public class EntitiesConverter<T> implements Converter<T> {
```
remove it, you don't need this anymore

dudarenko-dmitry replied:
I don't need all classes in package src/pl/senla/hotel/ie/file/archive/.
Deleted them all.

Mark005 reviewed (changes requested):
also, clean up everything, remove commented code if you don't need it and remove unused/odd classes

dudarenko-dmitry committed:
Task_7:
Delete unused/odd classes
